compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs

```diff
@@ -61,12 +61,26 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                     Place::ty_from(local, proj_base, self.body, self.infcx.tcx).ty
                 ));
 
-                item_msg = format!("`{}`", access_place_desc.unwrap());
-                if self.is_upvar_field_projection(access_place.as_ref()).is_some() {
-                    reason = ", as it is not declared as mutable".to_string();
+                // If the place is immutable then:
+                //
+                // - Either we deref a immutable ref to get to our final place.
+                //    - We don't capture derefs of raw ptrs
+                //    - If we derefed a mutable ref without deref-ing a immutable ref
+                //      to access this place then we will have mutability.
+                //    - So if we dereferenced something such that it resulted in the final
+                //      place being immutable, then we must have deref-ed a immut ref.
+                // - Or the final place is immut because the root variable of the capture
+                //   isn't marked mut and we should suggest that to the user.
+                if self.upvars[upvar_index.index()].place.place.deref_tys().next().is_none() {
```

Comment:
This is sort of a hack, I wasn't able to figure out why we are seeing extra reads. I have left a comment in the test below about what is getting read.